xlators/features/index/src/index.c

```diff
@@ -523,17 +523,22 @@ index_fill_readdir(fd_t *fd, index_fd_ctx_t *fctx, DIR *dir, off_t off,
             break;
         }
 
-        if (!strncmp(entry->d_name, XATTROP_SUBDIR "-",
-                     strlen(XATTROP_SUBDIR "-"))) {
-            check_delete_stale_index_file(this, entry->d_name, XATTROP_SUBDIR);
-            continue;
-        } else if (!strncmp(entry->d_name, DIRTY_SUBDIR "-",
-                            strlen(DIRTY_SUBDIR "-"))) {
-            check_delete_stale_index_file(this, entry->d_name, DIRTY_SUBDIR);
-            continue;
+        entry_dname_len = strlen(entry->d_name);
+        if (entry_dname_len >= 42) {
+            /* 42 = strlen('dirty-') + UUID_CANONICAL_FORM_LEN (36) */
+            if (!strncmp(entry->d_name, XATTROP_SUBDIR "-",
+                         strlen(XATTROP_SUBDIR "-"))) {
+                check_delete_stale_index_file(this, entry->d_name,
+                                              XATTROP_SUBDIR);
+                continue;
+            } else if (!strncmp(entry->d_name, DIRTY_SUBDIR "-",
+                                strlen(DIRTY_SUBDIR "-"))) {
+                check_delete_stale_index_file(this, entry->d_name,
+                                              DIRTY_SUBDIR);
+                continue;
+            }
```

Comment:
Actually the format of the name is very strict, it's just a fixed prefix, a `-`, and an uuid, so the length *must* be equal to 37 + length of prefix. If it's shorter or longer, it's not the desired entry.

We can define this function:
```c
static bool
index_check_index_file(const char *name, uint32_t nlen, const char *prefix)
{
    uint32_t plen;

    plen = strlen(prefix);
    if (nlen != plen + UUID_CANONICAL_FORM_LEN + 1) {
        return false;
    }

    if ((name[plen] != '-') || (memcmp(name, prefix, plen) != 0)) {
        return false;
    }

    check_delete_stale_index_file(this, name, prefix);

    return true;
}
```

And then rewrite this as:
```suggestion
        entry_dname_len = strlen(entry->d_name);
        if (index_check_index_file(entry->d_name, entry_dname_len,
                                   XATTROP_SUBDIR) ||
            index_check_index_file(entry->f_name, entry_dname_len,
                                   DIRTY_SUBDIR)) {
            continue;
        }
```

Reply:
I've done it slightly different, which I think is OK too - and cheap enough for the effort - comparison with either lengths and that's it.
(I do wonder if we should skip '.' and '..' as well, but that's in a differnet patch)